blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,701 @@
+
+Now let's take a closer look at some of the relevant components of the solution.
+
+=== The Outbox Table
+
+The `outbox` table resides in the database of the order service and has the following structure:
+
+[source]
+----
+Column        |          Type          | Modifiers
+--------------+------------------------+-----------
+id            | uuid                   | not null
+aggregatetype | character varying(255) | not null
+aggregateid   | character varying(255) | not null
+type          | character varying(255) | not null
+payload       | jsonb                  | not null
+----
+
+Its columns are these:
+
+* `id`: unique id of each message; can be used by consumers to detect any duplicate events, e.g. when restarting to read messages after a failure.
+Generated when creating a new event.
+* `aggregatetype`: the type of the aggregate root to which a given event is related; this could for instance be "purchase order" or "customer".
+This value will be used as the name of the corresponding topic in Kafka, so there'd be a topic for all events related to purchase orders,
+one topic for all customer-related events etc.
+Note that also events pertaining to a sub-entity contained within an aggregate should use that same type,
+so e.g. an event representing the cancelation of an individual order line should also use the aggregate type "order",
+ensuring that also this event will go into the "order" Kafka topic.
+* `aggregateid`: the id of the aggregate root that is affected by a given event; this could for instance be a purchase order id or a customer id;
+Similar to the aggregate type, events pertaining to a sub-entity contained within an aggregate should use the id of the containing aggregate root,
+e.g. the purchase order id for an order line cancelation event.
+This id will be used as the key for Kafka message later on.
+That way, all events pertaining to one aggregate root or any of its contained sub-entities will go into the same partition of that Kafka topic,
+which ensures that consumers of that topic will consume all the events related to one and the same aggregate root in the exact order as they were produced.
+* `type`: the type of event, e.g. "Order Created" or "Order Line Canceled". Allows consumers to trigger suitable event handlers.
+* `payload`: a JSON structure with the actual event contents, e.g. containing a purchase order, information about the purchaser, contained order lines, their price etc.
+
+=== Sending Events to the Outbox
+
+In order to "send" events to the outbox, code in the order service could in general just do an `INSERT` into the outbox table.
+However, it's a good idea to go for a slightly more abstract API, allowing to adjust implementation details of the outbox later on more easily, if needed.
+https://docs.jboss.org/weld/reference/latest/en-US/html/events.html[CDI events] come in very handy for this.
+They can be raised in the application code and will be processed _synchronously_ by the outbox event sender,
```

Comment:
May be worth mentioning that a Spring impl is quite similar. https://zoltanaltfatter.com/2017/06/09/publishing-domain-events-from-aggregate-roots/ and https://github.com/haithamshahin333/Domain-Event-Sample

Reply:
Added a reference to `@DomainEvents`.